jobs/populate_db.py

```diff
@@ -8,174 +13,148 @@
-            elif tool[field] != [] or tool[field] is not None:
-                completed_fields.append(field)
-                continue
-        elif tool["annotations"][field] == [] or tool["annotations"][field] is None:
-            empty_fields.append(field)
-        elif tool["annotations"][field] != [] or tool["annotations"][field] is not None:
-            completed_fields.append(field)
-    print({"Empty": empty_fields, "Completed": completed_fields})
-    add_tasks(empty_fields, tool_name)
-    remove_tasks(completed_fields, tool_name)
-
-
-def remove_tasks(fields, tool_name):
-    """
-    Receives fields and a tool name and removes incomplete tasks.
-
-    For each field in the list, the function removes a matching, INCOMPLETE (i.e., one with no value for "user") task from the db, if one exists.
-    """
-    for field in fields:
-        query = text(
-            "DELETE FROM task WHERE field_name = :field_name AND tool_name = :tool AND user IS NULL"
-        ).bindparams(field_name=field, tool=tool_name)
-        db.session.execute(query)
+
+
+def update_tool_table(tools: list[ToolhuntTool], timestamp, **kwargs) -> None:
+    """Upserts tool records and removes stale tools"""
+
+    [upsert_tool(tool, timestamp) for tool in tools]
+
+    remove_stale_tools()
+
+
+def insert_task(tool_name: str, field: str) -> None:
+    """Inserts a tool in the Tool table if it doesn't exist."""
+    query = text(
+        "SELECT * FROM task WHERE field_name = :field AND tool_name = :tool_name"
+    ).bindparams(field=field, tool_name=tool_name)
+    result = db.session.execute(query).all()
+    if len(result) == 0:
+        insert_stmt = insert(Task).values(
+            tool_name=tool_name,
+            field_name=field,
+        )
+        db.session.execute(insert_stmt)
         db.session.commit()
 
 
-def add_tasks(fields, tool_name):
-    """
-    Receives fields and a tool name and adds tasks.
-
-    For each field in the list, the function checks to see if a task exists for that particular tool and field type.  If none exists, it adds one.
-    """
-    for field in fields:
-        query = text(
-            "SELECT * FROM task WHERE field_name = :field_name AND tool_name = :tool"
-        ).bindparams(field_name=field, tool=tool_name)
-        result = db.session.execute(query).all()
-        if len(result) > 0:
-            print(f"A task for {tool_name}, {field} already exists in the database")
-            continue
-        else:
-            task = {"tool_name": tool_name, "field_name": field}
-            db.session.execute(insert(Task), task)
-            db.session.commit()
-            print(f"Added {field} task for {tool_name}")
+def update_task_table(tools: list[ToolhuntTool]) -> None:
+    """Inserts task records"""
+
+    for tool in tools:
+        tool_name = tool.name
+        for field in tool.missing_annotations:
+            insert_task(tool_name, field)
+
+
+# Pipeline
+
+
+def run_pipeline(**kwargs) -> None:
```

Comment:
It'd be nice to have some logging just inside this one function, to know where we are in the pipeline and that something is happening.

Reply:
Addressed with [04b56f0](https://github.com/wikimedia/toolhunt/pull/56/commits/04b56f03c36181953e85a5ba57bc744d5e8eeff3)